pkg/neonvm/ipam/ipam.go

```diff
@@ -42,32 +37,85 @@ const (
 	IpamRequestTimeout = 10 * time.Second
 
 	// DatastoreRetries defines how many retries are attempted when reading/updating the IP Pool
-	DatastoreRetries             = 5
-	DatastoreRetriesDelay        = 100 * time.Millisecond
-	DefaultLeaderLeaseDurationMs = 3000
-	DefaultLeaderRenewDeadlineMs = 2500
-	DefaultLeaderRetryPeriodMs   = 2000
+	DatastoreRetries      = 5
+	DatastoreRetriesDelay = 100 * time.Millisecond
+
+	// CleanupInterval defines how often the IPAM cleaner is run
+	CleanupInterval = 15 * time.Minute
 )
 
+var ErrAgain = errors.New("IPAM concurrency limit reached. Try again later.")
+
 type Temporary interface {
 	Temporary() bool
 }
 
 type IPAM struct {
 	Client
 	Config IPAMConfig
+
+	mu                 sync.Mutex
+	concurrencyLimiter *semaphore.Weighted
+}
+
+func (i *IPAM) AcquireIP(ctx context.Context, vmName types.NamespacedName) (net.IPNet, error) {
+	ip, err := i.runIPAM(ctx, getAcquireAction(vmName))
+	if err != nil {
+		return net.IPNet{}, fmt.Errorf("failed to acquire IP: %w", err)
+	}
+	return ip, nil
+}
+
+func (i *IPAM) ReleaseIP(ctx context.Context, vmName types.NamespacedName) (net.IPNet, error) {
+	ip, err := i.runIPAM(ctx, getReleaseAction(vmName))
+	if err != nil {
+		return net.IPNet{}, fmt.Errorf("failed to release IP: %w", err)
+	}
+	return ip, nil
+}
+
+func (i *IPAM) runCleanup(ctx context.Context, namespace string) error {
+	for {
+		vms, err := i.VMClient.NeonvmV1().VirtualMachines(namespace).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("error listing virtual machines: %w", err)
+		}
+		vmsMap := make(map[string]bool)
+		for _, vm := range vms.Items {
+			vmsMap[fmt.Sprintf("%s/%s", vm.Namespace, vm.Name)] = true
+		}
+		_, err = i.runIPAM(ctx, getCleanupAction(vmsMap))
```

Comment:
question: What happens if the set of VMs that have been assigned extra IPs changes between the call to `List()` and running the cleanup action?

Reply:
Yeah, that's fair. I have an idea for a fix, but I agree we should drop it.